### ISSUE TITLE
Receipt OCR

### ISSUE THREAD

**Issue by richardforrestbarker** [richardforrestbarker/Bardcode#27]
# Feature: OCR Receipts (Hybrid LayoutLMv3-based)

## Purpose
Implement a hybrid receipt OCR + structured-extraction pipeline:
- Client (Blazor): lightweight preprocessing (denoise/greyscale/deskew), upload images.
- Server: robust OCR → Layout-aware model (default: LayoutLM v3) for field extraction.
Make LayoutLMv3 the default but configurable to swap models (LayoutLMv2, Donut, smaller models) for testing/fallback.

## High-level architecture
- Blazor client (WASM/Server): image capture, client preprocessing, upload UI, manual correction fallback.
- API server: using the ApiService project. handles uploads and orchestrates pipeline through docker container CLI.
- OCR & Detector: PaddleOCR using PP-StructureV3.
- Layout model: HuggingFace / PyTorch running LayoutLMv3; model path/configurable.
- Storage: temporary object store for images (S3/MinIO) and metadata.
- Inference infra: GPU-backed instances for model inference; CPU fallback with degraded performance.

## Goals / Fields to extract
Initial schema:
- vendor_name
- merchant_address (optional)
- date
- total_amount
- subtotal
- tax_amount
- currency
- line_items: [{ description, quantity, unit_price, line_total, bounding_box }]
- raw_ocr_text (per page/image)
Per-field: confidence and bounding box references.

## Client-side preprocessing (Blazor)
Do simple, fast transforms to improve OCR:
- Convert to grayscale.
- Denoise (fast median or bilateral filter).
- Resize to target DPI (e.g., target height 1600 px, or 300 DPI equivalent).
- Auto-rotate / deskew.
- Contrast enhancement / adaptive thresholding.
- Optional perspective correction / crop detection.
- Optionally allow user to manually crop/rotate each image before upload.
- Upload images as multipart/form-data to server + metadata.

UX tips:
- Show upload & processing progress.
- Show extracted fields and allow correction on client before final save.
- Allow re-upload of single pages.

## Server-side pipeline (recommended steps)
1. Ingest
   - Accept multipart images, store temp object, generate job id.
2. Preprocess (reuse client normalization)
   - Additional denoise, morphological ops for low-quality images.
3. Text-detection & OCR
   - Run detector (integrated PaddleOCR detector) to get tight boxes.
   - Run recognizer (PaddleOCR) to get word-level text + boxes + confidences.
   - Fallback to Tesseract only when other OCR not available.
4. Tokenization & box mapping
   - Normalize boxes to model space (0–1000 for LayoutLM; preserve aspect ratio).
   - Tokenize words using model tokenizer; map each token → bounding box:
     - Default: assign the parent word’s box to all sub-tokens.
     - If tokenizer splits heavily, consider splitting box horizontally or keep parent box (simpler & commonly used).
5. Model input assembly
   - Build LayoutLMv3 inputs: token ids, attention masks, token boxes, and optionally image features.
   - For LayoutLMv3, include page image (resized) to compute visual embeddings.
6. Inference
   - Run LayoutLMv3 (fine-tuned for receipts if available).
   - Return entity spans, labels, and per-token/entity confidences.
7. Postprocessing
   - Convert entities to target schema (parse amounts, dates).
   - Merge multi-image document results; heuristics for split lines across pages.
   - Apply rules to verify totals (e.g., sum of line items ≈ subtotal).
8. Return result
   - JSON with structured fields, per-field confidences, and the raw OCR + coordinates.
   - Provide links to cropped images for manual review.

## API
- POST /api/receipts/upload
  - Body: multipart/form-data (files[]), metadata (merchant_id, timezone, user_id)
  - Response: 302-> /api/receipts/status/{job_id}
- GET /api/receipts/status/{job_id}
  - Response: { status: "processing" | "done" | "failed", progress, }
- GET /api/receipts/result/{job_id}
  - Response: {
      job_id,
      pages: [{ page_number, raw_ocr_text, words: [{text, box, conf}] }],
      fields: { vendor_name: {value, conf, box}, total_amount: {...}, ... },
      line_items: [{description, qty, unit_price, line_total, box, conf}]
    }

Box format:
- Normalized to model coordinate space: { x0, y0, x1, y1 } mapped to 0–1000 scale (ints).

## Model & configuration
- Default model: LayoutLMv3 (HuggingFace).
- Configurable options:
  - model_name_or_path (string)
  - device: auto-select GPU, fallback to CPU
  - ocr_engine: "paddle" | "tesseract"
  - detection_mode: "word" | "line"
  - box_normalization_scale: 1000 (default)
- Provide runtime toggle to switch to Donut/Image->structured if testing end-to-end.

## Token↔Box mapping (important)
- Input: OCR outputs words with boxes.
- Tokenize each word using the model tokenizer.
- Map each produced token to the word’s box (simple, robust).
- Ensure order of tokens matches reading order; order by top->bottom, left->right (with line grouping).

## Training & fine-tuning data
- Start with public datasets: CORD, SROIE (for receipts), FUNSD for forms patterns.
- Labeling: Label Studio / CVAT / doccano
- Data needed: several hundred to a few thousand receipts for decent results; more needed for many vendors/formats.
- Augmentation: synthesize receipts with varying fonts, noise, rotations.
- put the code in a new folder since it will be python. 

## Evaluation & metrics
- Field-level F1/Precision/Recall for vendor/date/total.
- Item-level accuracy (line item detection & totals).
- OCR word error rate (WER) and character error rate (CER).
- End-to-end extraction accuracy for real inputs.
- CI: unit tests for tokenizer→box mapping, integration tests with sample receipts.

## Deployment & infra
- Containerize the trained model in Docker using a command line interface to interact with the model 
- Use GPU-enabled instances (NVIDIA) for model inference; use CUDA + PyTorch.
- Use a queue (Redis + RQ / Celery) to process jobs and scale workers.
- reusing running model containers in memory across requests to reuse tokenizer/model objects.
- Consider autoscaling inference workers based on queue length.

## Privacy & security
- Encrypt images in transit (HTTPS) and at rest (server-side encryption).
- TTL for stored images; delete after result retrieval or after retention policies. 

## Failures, fallback & UX
- Low-confidence results → show manual correction UI.
- OCR failures → try alternate OCR engine (Tesseract fallback).
- If total parsing fails, surface raw OCR to the user with the highlight of likely total candidates.
- Log confidence and provide per-field confidence for automated accept/reject.

## Milestone features
- [ ] Fine-tune LayoutLMv3 on labeled receipts; iterate this one a few times building out datasets and fine-tuning. The model must be trimmed.
- [ ]  Image upload + simple OCR pipeline (PaddleOCR) + return raw OCR.
- [ ]  Implement token↔box mapping + LayoutLMv3 inference on server; simple field extraction.
- [ ] Add client preprocessing, UX for manual corrections, multi-page merge logic.
- [ ] Production hardening: rate-limiting, autoscaling, monitoring, privacy review.

## Risks & caveats
- LayoutLMv3 is large: requires GPU/VRAM; CPU inference is slow.
- Token-to-box mapping errors are a frequent source of extraction mistakes.
- OCR quality is mission-critical: prefer modern OCR (PaddleOCR) instead of relying on Tesseract for production.
- Limited labeled receipt diversity will limit model generalization.

## Acceptance criteria
- End-to-end pipeline processes multi-page receipts and extracts specified fields with >89.999999% field-level accuracy (set target after baseline).
- API responses follow schema above and include confidences and box data.
- Client allows user validation/correction of extracted results.
